File: CPP/ByRound/0963/0963_A_Alternating_Sum.cpp
// Problem: CF 963 A - Alternating Sum
// https://codeforces.com/contest/963/problem/A

#include <iostream>
#include <string>

using namespace std;
#define ll long long
const ll md = 1e9 + 9;
ll n, a, b, k, ans, sum, xa, xi, x, ia;
ll ksm(ll a, ll b) {
  ll re = 1;
  a %= md;
  while (b) {
    if (b & 1) {
      re = re * a % md;
    }
    a = a * a % md;
    b /= 2;
  }
  return re;
}
string s;
int main() {
  std::ios::sync_with_stdio(false);
  cin.tie(0);
  cout.tie(0);
  cin >> n >> a >> b >> k;
  cin >> s;
  ans = 0;
  x = ksm(a, n);  // x = a^n
  ia = ksm(a, md - 2), xa = 1;  // ia = a^(-1) mod md, xa = 1
  for (int i = 0; i < k; i++) {
    if (s[i] == '+') {
      ans = (ans + x * xa % md) % md;  // Add contribution of s[i] = +1
    } else {
      ans = (ans + md - x * xa % md) % md;  // Add contribution of s[i] = -1
    }
    x = x * ia % md, xa = xa * b % md;  // Update x for next term and xa for next coefficient
  }
  xa = 1;
  sum = 0;
  xi = ksm(b * ia, k);  // xi = (b * a^(-1))^k
  x = (n + 1) / k;  // Number of complete cycles
  if (xi == 1) {
    cout << ans * x % md << endl;  // If the sum is 1, just multiply by number of cycles
  } else {
    ll xxx = (1 + md - ksm(xi, x));  // Sum of geometric series: (1 - xi^x)
    ll xxx1 = ksm(1 + md - xi, md - 2);  // Inverse of (1 - xi)
    cout << ans * xxx % md * xxx1 % md << endl;  // Final result is ans * (1 - xi^x) / (1 - xi)
  }
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/0963/0963_A_Alternating_Sum.py
# Problem: CF 963 A - Alternating Sum
# https://codeforces.com/contest/963/problem/A

# '''
# Problem: Alternating Sum
# 
# Algorithm:
# This problem requires computing a sum of the form:
# sum_{i=0}^{n} s_i * a^{n-i} * b^i
# where s_i is a k-periodic sequence of +1 and -1.
# 
# Key Observations:
# 1. The sequence s is k-periodic, so we only need to store the first k elements.
# 2. The terms of the sum can be rewritten as:
#    a^n * (s_0 * (b/a)^0 + s_1 * (b/a)^1 + ... + s_{k-1} * (b/a)^{k-1})
#    where (b/a) is used in geometric progression.
# 
# Approach:
# - Compute the contribution of one full period of the sequence.
# - Use geometric progression sum formula to compute total sum efficiently.
# - Modular arithmetic is used throughout to avoid overflow.
# 
# Time Complexity: O(k + log(n)) due to power computations and one loop over k elements.
# Space Complexity: O(k) for storing the sequence.
# '''

M = 10**9 + 9
n, a, b, k = [int(e) for e in input().split()]
seq = [1 if e == "+" else -1 for e in input()]
s = 0
x = pow(a, n, M) # Precompute a^n mod M
y = (pow(a, -1, M) * b) % M # Compute (b/a) mod M
for i in range(k):
    s += x * seq[i] # Add contribution of current term
    x *= y # Multiply by (b/a) for next term in the sum
    s %= M # Keep s in mod
    x %= M # Keep x in mod
q = pow(y, k, M) # Compute (b/a)^k mod M
S = s * ((n + 1) // k) # Total number of full periods
if q != 1:
    # Apply geometric series formula for sum of a^k + a^{2k} + ... + a^{(n+1)/k * k}
    # which is a^k * (1 - a^{(n+1)/k * k}) / (1 - a^k)
    S = s * (pow(q, (n + 1) // k, M) - 1) * pow(q - 1, -1, M) # Apply geometric progression sum
print(S % M)


# https://github.com/VaHiX/CodeForces/